app/services/request_password_reset.rb

```diff
@@ -23,11 +23,31 @@ def instructions
     I18n.t('user_mailer.email_confirmation_instructions.first_sentence.forgot_password')
   end
 
-  def user_not_found?
-    user.is_a?(NonexistentUser)
+  ##
+  # If a user record does not exist for an email address, we send a registration
+  # email instead of a reset email so the user can go through the account
+  # creation process without having to receive another email
+  #
+  # If a user exists but does not have any confirmed email addresses, we send
+  # them a reset email so they can set the password on the account
+  #
+  # If a user exists and has a confirmed email addresses, but this email address
+  # is not confirmed we should not let them reset the password with this email
+  # address. Instead we send them an email to create an account with the
+  # unconfirmed email address
+  ##
+  def user_should_receive_registration_email?
+    return true if user.is_a?(NonexistentUser)
```

Comment:
IMO the purpose of the null object pattern is to avoid type checks like this, WDYT of adding `.exists?` to `User` and `NonexistentUser` classes and using that here?

Reply:
Yeah. When I was looking at this it looked like this is really the last place the `NonexistentUser` user gets used. I see an opportunity to get rid of it entirely. Let me go for that and see if it works out.

Reply:
That works too!